### ISSUE TITLE
PEP 562 support

### ISSUE THREAD

**Issue by dumbPy** [microsoft/pyright#465]
**Is your feature request related to a problem? Please describe.**
[PEP 562](https://www.python.org/dev/peps/pep-0562/) introduces `__getattr__` and `__dir__` for modules (and also works for packages in their `__init__.py`). The PEP documentation explains why it's needed and has been introduced in python 3.7+

**Describe the solution you'd like**
pyright should recognise the attributes from `__dir__` of a module or a package's `__init__.py`. Ofcourse the static type cannot be determined as it is not included in PEP 562

**Additional context**
simple testcase
```
.
├── module_foo.py
├── package_foo
│  └── __init__.py
└── run.py
```
```python
# module_foo.py and package_foo/__init__.py
def __dir__(): return ["bar", "baz"]
_foo_bar_baz = True
def __getattr__(name):
    if name in __dir__():
        return _foo_bar_baz

# run.py
from module_foo import bar # <-- pyright raises "'bar' is unknown import symbol"
from package_foo import baz # <-- pyright raises "'bar' is unknown import symbol"

print(bar)
print(baz)

```
```zsh
$ python run.py
True
True
```

**Comment by erictraut** [comment]
A static type checker wouldn't be able to evaluate the return value of `__dir__`. In general, the function could contain conditional statements, loops, function calls, etc. 

The best Pyright could do in this case is assume that any imported symbol from a module that has a `__getattr__` method is valid but of unknown type. That would eliminate the error but also mask other potential errors. I guess it's similar to a class that defines a `__getattr__` method.

The best approach in this case is to create a type stub file that properly describes the interface to the module, including any lazily-initialized classes. Pyright will then use the type stub rather than the implementation to resolve the import symbols.

**Comment by dumbPy** [comment]
In that case, stub file seems like a good idea rather than assuming everything from a module to be a valid attribute.  
May be a warning suggesting to implement a stub file everytime there's a `__getattr__`   method would be a good idea though.  

Thanks again.  
Pyright is the way  :smile_cat:

**Comment by erictraut** [comment]
I think it's still worth adding support for module-level `__getattr__` functions. Pyright should treat them the same as class-level `__getattr__` methods.

**Comment by erictraut** [comment]
I just added the support along with some unit tests.
https://github.com/microsoft/pyright/commit/a25e832ad46b47d09fd9ccde77280e898bf9fc76

This will be in the next published version of Pyright.

Thanks again for the suggestion.

**Comment by erictraut** [comment]
This support is included in Pyright v1.1.17, which I just published.

**Comment by nathanrpage97** [comment]
Could we allow ``__getattr__`` to work with typings?


_foo.py_
```python

def __getattr__(name: str) -> int:
    ...
``` 

_main_
```python
import foo
from foo import a # a would be an int instead of unknown

b = foo.b # this would indicate b is an int.
```

**Comment by erictraut** [comment]
Thanks for the suggestion. This will be implemented in the next release of pyright.

**Comment by erictraut** [comment]
This is now addressed in Pyright 1.1.98, which I just published. It will also be in the next release of Pylance.
